File: src/BinarySearch/lc0704_BinarySearch.py
# https://leetcode.cn/problems/binary-search
class Solution:
    def search(self, nums: list[int], target: int) -> int:
        if len(nums) == 0:
            return -1

        def first_greater_than_or_equal_to(lo: int, hi: int) -> int:
            while True:
                if lo == hi:
                    return lo
                mid = (lo + hi) // 2
                if nums[mid] >= target:
                    hi = mid
                else:
                    lo = mid + 1

        possible_index = first_greater_than_or_equal_to(0, len(nums))
        if possible_index <= len(nums) - 1:
            if nums[possible_index] == target:
                return possible_index
            else:
                return -1
        else:  # possible_index == len(nums)
            return -1

File: src/BinarySearch/lc0704_BinarySearch.go
package BinarySearch

// https://leetcode.cn/problems/binary-search
func search(nums []int, target int) int {
	if len(nums) == 0 {
		return -1
	}
	leftIndex := 0
	rightIndex := len(nums) - 1
	for {
		if rightIndex < leftIndex {
			return -1
		}
		midIndex := (leftIndex + rightIndex) / 2
		if nums[midIndex] > target {
			rightIndex = midIndex - 1
		} else if nums[midIndex] < target {
			leftIndex = midIndex + 1
		} else { // leftIndex == rightIndex
			return midIndex
		}
	}
}
